api/src/main/java/com/flipkart/poseidon/validator/TaskValidator.java

```diff
@@ -34,14 +40,35 @@
         for (Map.Entry<String, TaskPOJO> entry : tasks.entrySet()) {
             String taskName = entry.getKey();
             TaskPOJO task = entry.getValue();
+            final Class<? extends DataSource<?>> dsClass = datasources.get(task.getName());
             if (isNullOrEmpty(task.getName())) {
                 errors.add("No datasource defined for Task: " + braced(taskName));
-            } else if (datasources.get(task.getName()) == null && validateDataSources) {
+            } else if (dsClass == null && validateDataSources) {
                 errors.add("Datasource used does not exist for Task: " + braced(taskName));
             }
 
-            Map<String, Object> context = Optional.ofNullable(task.getContext()).orElse(new HashMap<>());
+            final Set<String> datasourceRequestAttributes = new HashSet<>();
+            if (dsClass != null && validateDataSources) {
+                final List<? extends Constructor<?>> injectableConstructors = findInjectableConstructors(dsClass);
+                if (!injectableConstructors.isEmpty()) {
+                    final Constructor<?> constructor = injectableConstructors.get(0);
+                    final Parameter[] parameters = constructor.getParameters();
+                    for (Parameter parameter : parameters) {
+                        final RequestAttribute requestAttribute = parameter.getAnnotation(RequestAttribute.class);
+                        if (requestAttribute != null) {
+                            datasourceRequestAttributes.add(Optional.of(requestAttribute.value()).filter(StringUtils::isNotEmpty).orElse(parameter.getName()));
+                        }
+                    }
+                }
+            }
+
+            final Map<String, Object> context = Optional.ofNullable(task.getContext()).orElse(new HashMap<>());
             for (Map.Entry<String, Object> contextEntry : context.entrySet()) {
+                final String key = contextEntry.getKey();
+                if (!datasourceRequestAttributes.contains(key)) {
+                    errors.add("ContextParam: " + braced(key) + " used in Task: " + braced(taskName) + " is not used in the Datasource");
```

Comment:
Hope reverse condition is already handled

Reply:
reverse can't really be verified as the DS might expect some nullable fields from other sources